File: Rococo's Image2PDF/FSSettingsViewController.m
//
//  FSSettingsViewController.m
//  Rococo's Image2PDF
//
//  Created by Zhefu Wang on 13-6-24.
//  Copyright (c) 2013年 Nonomori. All rights reserved.
//

#import "FSSettingsViewController.h"
#import "FSNavigationController.h"
#import "UIViewController+FSViewController.h"
#import "FSPagingScrollView.h"
#import "FSPageValueDefinition.h"
#import "JSONKit.h"
#import "UIButton+FSUIButton.h"

@interface FSSettingsViewController ()

@property (weak, nonatomic) IBOutlet UIView *contentView;
@property (weak, nonatomic) IBOutlet FSPagingScrollView *scrollView;
@property (nonatomic, strong) NSArray *pageArray;
@property (nonatomic) NSInteger pageIndex;
@property (weak, nonatomic) IBOutlet UILabel *descriptionLabel;
@property (nonatomic, strong) NSDictionary *allPageDictionary;

@end

@implementation FSSettingsViewController

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.pageArray = @[@"LETTER",@"A3",@"A4",@"A5",@"B4",@"B5"];
    self.scrollView.contentSize = self.contentView.frame.size;
    [self addObserver:self forKeyPath:@"pageIndex" options:NSKeyValueObservingOptionOld context:nil];
    NSData *json = [NSData dataWithContentsOfFile:[[NSBundle mainBundle] pathForResource:@"pageSize" ofType:@"json"]];
    self.allPageDictionary = [[[JSONDecoder alloc] init] objectWithData:json];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    for (int i = 0; i < [self.pageArray count]; ++i)
    {
        NSDictionary *pageDic = self.allPageDictionary[[NSString stringWithFormat:@"PAGESIZE_%@",self.pageArray[i]]];
        CGRect frame;
        frame.origin.x = self.scrollView.frame.size.width * i + 20;
        frame.origin.y = 12;
        frame.size = CGSizeMake(160, 200);
        UIImageView *subview = [[UIImageView alloc] initWithFrame:frame];
        subview.image = [UIImage imageNamed:pageDic[@"image"]];
        [self.scrollView addSubview:subview];
    }
    self.scrollView.contentSize = CGSizeMake(200 * [self.pageArray count], self.scrollView.frame.size.height);
    self.scrollView.responseInsets = UIEdgeInsetsMake(0, 60, 0, 60);
    NSDictionary *pageSizeDictionary = [[NSUserDefaults standardUserDefaults] objectForKey:@"pageSize"];
    CGRect rect = CGRectMake(200*([self.pageArray indexOfObject:pageSizeDictionary[@"name"]]+1), 0, 1, 1);
    [self.scrollView scrollRectToVisible:rect animated:YES];
}

- (void)dealloc
{
    [self removeObserver:self forKeyPath:@"pageIndex"];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)cancelButtonTouched:(UIButton *)sender
{
    FSNavigationController *navCV = (FSNavigationController*)self.navigationController;
    [navCV popViewControllerWithSlideAnimation];
}

- (BOOL)isAllowedSwipeBack
{
    return NO;
}

- (IBAction)commitButtonTouched:(UIButton *)sender
{
    __weak __block FSSettingsViewController *weakSelf = self;
    [UIView animateWithDuration:0.05
                     animations:^{
                         sender.frame = sender.originalFrame;
                     }
                     completion:^(BOOL finished){
                         if (finished)
                             [sender touched:^{
                                 
                                 NSDictionary *pageSizeDic = self.allPageDictionary[[NSString stringWithFormat:@"PAGESIZE_%@",weakSelf.pageArray[weakSelf.pageIndex]]];
                                 NSLog(@"%@",pageSizeDic);
                                 [[NSUserDefaults standardUserDefaults] setObject:pageSizeDic forKey:@"pageSize"];
                                 [[NSUserDefaults standardUserDefaults] synchronize];
                                 FSNavigationController *navCV = (FSNavigationController*)self.navigationController;
                                 [navCV popViewControllerWithSlideAnimation];
                             }];
                     }];    
}

- (IBAction)okButtonTouchCancel:(UIButton *)sender
{
    [UIView animateWithDuration:0.1 delay:0 options:UIViewAnimationOptionCurveEaseIn animations:^{
        sender.frame = sender.originalFrame;
    } completion:nil];
}

- (IBAction)okButtonTouchDown:(UIButton *)sender
{
    sender.originalFrame = sender.frame;
    [UIView animateWithDuration:0.1
                          delay:0
                        options:UIViewAnimationOptionCurveEaseIn
                     animations:^
     {
         CGRect frame = sender.frame;
         CGFloat increasedWidth = frame.size.width*0.3;
         CGFloat increasedHeight = frame.size.height*0.3;
         frame.origin.x -= increasedWidth;
         frame.origin.y -= increasedHeight;
         frame.size.width += increasedWidth*2;
         frame.size.height += increasedHeight*2;
         sender.frame = frame;
     }
                     completion:nil];
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
    if ([keyPath isEqualToString:@"pageIndex"])
    {        
        NSDictionary *pageSizeDic = self.allPageDictionary[[NSString stringWithFormat:@"PAGESIZE_%@",self.pageArray[self.pageIndex]]];

        self.descriptionLabel.text = [NSString stringWithFormat:@"%.2fx%.2f inch",[pageSizeDic[@"width_inch"] floatValue],[pageSizeDic[@"height_inch"] floatValue]];
    }
}

#pragma mark - UIScrollViewDelegation
- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    CGFloat pageWidth = self.scrollView.frame.size.width;
    self.pageIndex = floor((self.scrollView.contentOffset.x-pageWidth/2)/pageWidth) + 1;
}

@end

File: Rococo's Image2PDF/FSSettingsViewController.swift
 //
//  FSSettingsViewController.swift
//  Rococo's Image2PDF
//
//  Created by zhefu wang on 10/7/14.
//  Copyright (c) 2014 Nonomori. All rights reserved.
//

import UIKit

class FSSettingsViewController: UIViewController, UIScrollViewDelegate {
    @IBOutlet weak var scrollView: FSPagingScrollView!
    @IBOutlet weak var descriptionLabel: UILabel!
    let pageArray: [String] = ["LETTER","A3","A4","A5","B4","B5"]
    var pageIndex: Int = 0 {
        didSet{
            guard let pageSizeDict: [String:AnyObject] = allPageDictionary?["PAGESIZE_\(pageArray[pageIndex])"] else {
                return
            }
            guard let width: Float = pageSizeDict["width_inch"] as? Float, height: Float = pageSizeDict["height_inch"] as? Float else {
                return
            }
            descriptionLabel.text = String(format: "%.2fx%.2f inch", width, height)
        }
    }
    
    var allPageDictionary: [String:[String:AnyObject]]?
    
    override func viewDidLoad() {
        super.viewDidLoad()
       
        let jsonData: NSData = try! NSData(contentsOfFile: NSBundle.mainBundle().pathForResource("pageSize", ofType: "json")!, options: [])
        
        allPageDictionary = (try? NSJSONSerialization.JSONObjectWithData(jsonData, options: .MutableContainers)) as? [String:[String:AnyObject]]
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        
        for i in 0..<pageArray.count {
            var pageDic: Dictionary<String, AnyObject>? = allPageDictionary!["PAGESIZE_\(pageArray[i])"]
            var frame:CGRect = CGRectZero
            frame.origin.x = scrollView.frame.size.width * CGFloat(i) + 20
            frame.origin.y = 12
            frame.size = CGSize(width: 160, height: 200)
            let subview = UIImageView(frame: frame)
            
            if let imageName = pageDic?["image"] as? String{
                subview.image = UIImage(named: imageName)
            }
            scrollView.addSubview(subview)
            scrollView.contentSize = CGSize(width: CGFloat(200 * pageArray.count), height: scrollView.frame.size.height)
            scrollView.responseInsets = UIEdgeInsets(top: 0, left: CGRectGetMidX(view.bounds) - CGRectGetMidX(scrollView.bounds), bottom: 0, right: CGRectGetMidX(view.bounds) - CGRectGetMidX(scrollView.bounds))
            
            if let pageSizeDict = NSUserDefaults.standardUserDefaults().objectForKey("pageSize") as? Dictionary<String, AnyObject>{
                if let name = pageSizeDict["name"] as? String{
                    if let index = pageArray.indexOf(name){
                        let rect = CGRect(x: CGFloat(200 * index), y: 0, width: scrollView.frame.size.width, height: scrollView.frame.size.height)
                        scrollView.scrollRectToVisible(rect, animated: true)
                    }
                }
            }
        }
    }
    
    
    @IBAction func cancelButtonTouched(sender: UIButton) {
        guard let pageSizeDict: [String:AnyObject] = allPageDictionary?["PAGESIZE_\(pageArray[pageIndex])"] else {
            return
        }
        NSUserDefaults.standardUserDefaults().setObject(pageSizeDict, forKey: "pageSize")
        NSUserDefaults.standardUserDefaults().synchronize()
        if let navCV: FSNavigationController = navigationController as? FSNavigationController{
            navCV.popViewControllerWithSlideAnimation()
        }
    }
    
    func isAllowedSwipeBack() -> Bool{
        return false
    }
    
    //MARK: - UIScrollViewDelegate
    func scrollViewDidScroll(scrollView: UIScrollView) {
        let pageWidth = scrollView.frame.size.width
        var tmp = Int(floor((scrollView.contentOffset.x - pageWidth * 0.5)/pageWidth)) + 1
        if tmp < 0 {
            tmp = 0
        }
        if tmp >= pageArray.count {
            tmp = pageArray.count - 1
        }
        pageIndex = tmp
    }

}
